<a href="https://colab.research.google.com/github/sreekanthreddysri2017/Data_Science/blob/main/5_01_Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b>Feature Engineering</b>

## **Agenda**

 we will cover the following concepts with the help of a business use case:
* Features
* Feature engineering and it's methods
   * Applying domain expertise
   * Combining features
   * Encoding of categorical variables

## **Features**

* In most cases, machine learning algorithms are fed a collection of numeric examples that are stacked on top of each other to form a two-dimensional feature matrix.

* This matrix has one **example** per row and one **function** per column.

* Understanding the task of these features is critical when doing some kind of learning on the dataset.

* Since the algorithms create their own internal transformations, deep learning features are normally simple.

* This method necessitates a vast volume of data and sacrifices interpretability; but, in image processing and natural language processing applications, these trade offs are often worthwhile.

* Feature engineering is needed for the majority of other use cases in order to translate data into a machine learning-ready format.

* The features chosen are critical for both the data's interpretability and the model's performance.

* A precise machine learning framework cannot be built without feature engineering.

## <b> What Is Feature Engineering?</b>

* Feature Engineering repurposes already modified features to build new ones, making it possible for any Machine Learning algorithm to comprehend and master any pattern.

* It is the process of **creating new features** from raw data using domain knowledge to make the machine learning algorithms work.

* It is one of the fundamental concepts of machine learning and it is both difficult and expensive.

* Every new resource must enhance the performance in some way; otherwise, it would have the opposite effect, worsening the final outcome. When this happens, we must use feature selection instead of feature engineering.

* Acording to Andrew Ng, "Coming up with features is difficult, time-consuming, requires expert knowledge."

* According to Pedro Domingos, feature engineering is the art of coming up with new features with more predictive power using:

   - Experience
   - Domain expertise
   - Empirical processes

* The following are some of the most important components of feature engineering:
  1. Imputation
  2. Handling Outliers
  3. Binning
  4. One-Hot Encoding
  5. Grouping Operations
  6. Scaling

### Most Important Techniques in Feature Engineering

- Applying domain expertise
- Combining features
- Encoding of categorical variables

### <b> Applying Domain Expertise </b>

Domain expertise or domain knowledge is nothing but expertise in a particular field, such as Education, Healthcare, Consumer Goods, and Retail.
A domain expert is someone who is not related to the technology aspect but has in-depth knowledge about the particular industry, how it is shaping up, the trends, and the things that might impact the industry.

For example, you are called in to develop a particular application for a consumer goods company, specifically an apparel and footwear one.
The application that you build has to align to the industry and its various facets, and you being a technology expert wouldn't know much about it. This is where a domain expert will come in and explain how that industry works and what would be the best way to have the application built.

### <b> Combining Features </b>

This is one of the most crucial part of feature engineering that involves analyzing the data set and understanding the variable(s) to combine one or more features to create new ones. This process involves intuitive decision-making and a small amount of research about the domain of the data set.

Combining features helps to increase the performance of the model which is fitted on the data.


## <b>Encoding of Categorical Variables</b>

Let's first see what encoding is.

#### **What Is Encoding?**

Encoding is used to transform the categorical variable into numerical features.

For example, we have an attribute called gender in a data set where values are male and female. In this case, the numerical encoded version of the values will be 1 for male, 0 for female, or vice versa.


Since different kinds of categorical variables capture different amount of information, we need different techniques to encode them.

### **Label Encoding**

Label encoding is a handy technique to encode categorical variables. However, such encoded nominal variables might end up being misinterpreted as ordinal.

#### **Encoding Can Be Performed on the Following Types of Variables:**

#### Nominal Variables

Consider the following three categorical variables and their values:

**Color**: Blue, Green, Red, Yellow

**Educational Qualification**: Primary School, Secondary School, Graduate, Post-Graduate, PhD

**Salary Bracket**: 0-50,000, 50,001-100,000, 100,001-150,000, 150,001-200,000

Although all three of them are categorical variables, they are different in the amount of information they convey.
Let's look at them one by one.

Color conveys blue is different from red. That's all. The value of the variable is not meant to capture any relative difference among the values. Such variables are called `Nominal Variables`.

####  Ordinal Variables

Now consider educational qualifications.

The value "graduate" does not only convey that it is different from the value say "Primary school", it also implies that it is more in terms of qualification than "Primary School".

Such variables are called `Ordinal Variables` because they convey a sense of order.

In our example, Primary School < Secondary School < Graduate < Post-Graduate < PhD. (in terms of qualification).

#### Interval Variables

The third variable, salary bracket is similar to educational qualification by conveying order (a person earning 50,001-100,000 earns more salary than 0-50,000).

However, here, apart from knowing the order, we also know the interval between the values.

Here we can say that the averages of each of the values are separated by 50,000. Such variables are called `Interval variables`.

## Techniques Used for Encoding Variables
There are two types of broadly used algorithms which perform the task of encoding of variables.


### There are few libraries required to perform encoding variables:
* Pandas - It helps to retrieve datasets, handle missing data and perform data wrangling.
* NumPy - It helps to perform numerical operations in the dataset.
* sklearn.preprocessing - It helps in data transformation.

In [ ]:
#Select the cell and click on run icon to import libraries.

import pandas as pd
import numpy as np

# Import label encoder
from sklearn import preprocessing

**Instruction:**

 Download the **`Iris.csv`** dataset file from Kaggle

**Note 1:**  
* The **`iris_df`** is a dataframe that stores data imported from a CSV file in rows and columns format.
* The **head()** function helps to view the first few data present in the **`iris_df`** dataframe.

In [ ]:
#Select the cell and click on run icon to retrieve and view the dataset.
# Import dataset

iris_df = pd.read_csv('Iris.csv')
iris_df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

**Observations from the above output:**
>View few records of Iris dataset


**Note 2:**  
The **info()** function helps to understand the dataset, the column name, total null values, and data type.

In [ ]:
#Select the cell and click on run icon
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


**Observations from the above output:**
>The variable **`species`** is a categorical Dtype **'object'**.

**Note 3:**

The **LabelEncoder()** function is used to convert categorical variables into numerical. According to our dataset, variable **`species`** is categorical, so convert **`species`** into numerical type.

In [ ]:
#Select the cell and click on run icon to define LabelEnocder
label_encoder = preprocessing.LabelEncoder()

**Note 4:**

The **unique()** function is used to identify distinct rows present in the **`iris_df`** dataframe.

In [ ]:
#Select the cell and click on run icon
iris_df['species'].unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

**Observations from the above output:**
>There are three categories of variable **`species`** such as *'setosa'*, *'versicolor'*, and *'virginica'* to be encoded.

**Note 5:**

* The **fit_transform()** method calculates the mean and variance of each feature and transforms all the features using the respective mean and variance.
* The **head()** function helps to view the first few data present in the **`iris_df`** dataframe.

In [ ]:
#Select the cell and click on run icon
iris_df['species']= label_encoder.fit_transform(iris_df['species'])
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width  species
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

**Observations from the above output:**
>The variable **`species`** is converted into numerical variable as 0's and 1's.

In [ ]:
#Select the cell and click on run icon

iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


**Observations from the above output:**
>The variable **`species`** Dtype is converted to **'int64'**.

>>Using LabelEnocder, we can convert a categorical variable into a numerical variable.

## **One Hot Encoding**

A data set with more dimensions requires more parameters for the model to understand, and that means more rows to reliably learn those parameters.

The effect of using One Hot Encoder is the addition of a number of columns (dimensions).

If the number of rows in the data set is fixed, the addition of extra dimensions without adding more information for the models to learn from can have a detrimental effect on the eventual model accuracy.

If the number of rows in the dataset is fixed and the addition of extra dimension without adding more information in the model would affect the model accuracy.

In the below example, you can see that category of Variable X is enocded into separate columns such as Variable X_Blue, Variable X_Yellow, Variable X_Red.

![one_hot](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/Applied_Machine_Learning/Images/0.4_Feature_Engineering/Trainer_PPT_and_IPYNB/one_hot.png)

There are few libraries required to perform **One Hot Encoding**:

* **datasets**: It helps to load the datasets from sklearn library.

* **OneHotEncoder**: It helps to encode categorical variable.

In [ ]:
#Select the cell and click on run icon
from sklearn import datasets
from sklearn.preprocessing import OneHotEncoder

**Note 6:**

The given code helps to load the **Iris dataset** and create a dataframe **iris_data**, and **y** variable consists of target values.

In [ ]:
#Select the cell and click on run icon
iris_data = datasets.load_iris()

iris_data = pd.DataFrame(data=np.c_[iris_data["data"], iris_data["target"]], columns=iris_data["feature_names"] + ["target"])
y = iris_data.target.values

**Note 7:**

The **head()** function is used to view the first few data present in the **`iris_data`** dataframe.

In [ ]:
#Select the cell and click on run icon
iris_data.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0     0.0  
1     0.0  
2     0.0  
3     0.0  
4     0.0

**Observations from the above output:**
>There are four dependent variable and one independent variable present in the **`iris_data`** dataframe.


In [ ]:
# Select the cell and click on run icon to view the target values
y

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.])

**Observations from the above output:**
>The variable **`target`** is converted into an array.

**Note 8:**

* In OneHotEncoder, the categories set as 'auto' determines the categories automatically.
* The **reshape()** function reshapes an array without modifying its data.
* The **toarray()** function helps to represent the data in array format.

In [ ]:
#Select the cell and click on run icon
onehotencoder = OneHotEncoder(categories='auto')
y = onehotencoder.fit_transform(y.reshape(-1,1))
print(y.toarray())

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0.

**Observations from the above output:**
>With the help of above output, you can see that variable **`target`** is converted into numerical variable as 0's and 1's into three separate categories .

**Note 9:**

The **get_dummies()** function is used for data manipulation and converts categorical variable to numerical variables .

In [ ]:
#Select the cell and click on run icon
pd.get_dummies(iris_data.target).head()

0.0  1.0  2.0
0    1    0    0
1    1    0    0
2    1    0    0
3    1    0    0
4    1    0    0

**Observations from the above output:**
>The target values are converted into dummy values in three different columns.

## <b>Hashing</b>

Hashing is a technique that combines more than one category of a categorical variable into one single category.

There are three advantages in one-hot encoding or set-of word model:

- It’s much simpler to code
- It doesn’t require any dictionaries to prepare
- It’s friendly to online learning where you can train on a data set that doesn’t fit in memory because you need to see each example only once. One-hot encoding will not work well with online learning because to prepare dictionaries you need to see the whole data set first

## <b>Feature Hashing</b>


Feature hasing is a important technique for handling sparse and high-dimensional features in machine learning.

- It is fast, simple, memory-efficient, and well-suited to online learning sceanrios.
- It converts unique tokens into integers.
- It operates on the exact strings that you provide as input and does not perform any linguistic analysis or preprocessing.


## Techniques to merge categories:

### **Business Logic**

Consider a column in the dataset corresponds to "zip codes". There are 182 zip codes in New York state and it is impractical to use each zip code as a separate category.
So, to tackle this situation we can merge the zip codes according to localities.
This helps to reduce the number of categories and results in meaningful aggregation of zip code.

### **Frequency**

- It is not possible to apply business logic every time. In such cases, perform hashing using the frequency of occurrence.
- To combine levels using their frequency, we first look at the frequency distribution of each level and combine levels having frequency say less than 5% of total observation (can be changed based on distribution).
- This is an effective method to deal with rare levels.
- We can also combine levels by considering the response rate of each level. We can simply combine levels having similar response rates into the same group.

Now we'll see all the concepts explained with the help of a business use case.

### There is a libraries required to perform encoding variables:
* FeatureHasher - Feature hashing can be employed in document classification and helps to extract important features from large text corpus.


In [ ]:
#Select the cell and click on run icon
from sklearn.feature_extraction import FeatureHasher

**Instruction:**

Download the **`vgsales.csv`** dataset file from Course Resources and upload it in the lab using the Up arrow as shown below on the View Tab

**Note 10:**  
* The **`game_df`** is dataframe to store the data imported from the csv as rows and columns table format.
* The encoding is specified as 'UTF-8' as the file type converion of the csv file.
* The **head()** function helps to view first few data present in the dataframe.

In [ ]:
#Select the cell and click on run icon
game_df = pd.read_csv("vgsales.csv", encoding="utf-8")
game_df.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

**Note 11:**

The **columns** provides the name of the columns present in the **`game_df`** dataframe.

In [ ]:
#Select the cell and click on run icon
game_df.columns

Index(['Rank', 'Name', 'Platform', 'Year', 'Genre', 'Publisher', 'NA_Sales',
       'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales'],
      dtype='object')

**Observations from the above output:**
>The column names such as Rank, Name, Platform, Year, Genre, Publisher, NA_Sales,
       EU_Sales, JP_Sales, Other_Sales, and Global_Sales present in the **`game_df`** dataframe.

**Note 12:**

The **shape** represents the number of elements in each dimension and returns a tuple with each index having the number of corresponding elements.

In [ ]:
#Select the cell and click on run icon
game_df.shape

(16598, 11)

**Observations we can draw from the above output:**

>There are 16598 rows and 11 columns present in the **`game_df`** dataframe.

**Note 13:**

* The **iloc** is integer-index based that helps to extract the required number of rows or columns by specifying the integer index value present in the dataframe.
* Example: iloc[1:7] represents six rows, as it substracts one from seven.

In [ ]:
#Select the cell and click on run icon
game_df[['Name', 'Platform', 'Year', 'Genre', 'Publisher']].iloc[1:7]

Name Platform    Year         Genre Publisher
1         Super Mario Bros.      NES  1985.0      Platform  Nintendo
2            Mario Kart Wii      Wii  2008.0        Racing  Nintendo
3         Wii Sports Resort      Wii  2009.0        Sports  Nintendo
4  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo
5                    Tetris       GB  1989.0        Puzzle  Nintendo
6     New Super Mario Bros.       DS  2006.0      Platform  Nintendo

**Observations we can draw from the above output:**
>There are six rows present in the **`game_df`** dataframe as specified in the iloc.

**Note 14:**

* The **unique()** function extracts unique values present in a particular variable or dataframe.
* The **len()** function calculates the total number of elements present of a variable or dataframe.
* The **print()** function is used to display the result.

In [ ]:
#Select the cell and click on run icon
u_generes = np.unique(game_df["Genre"])
print("Total game generes:", len(u_generes))
print(u_generes)

Total game generes: 12
['Action' 'Adventure' 'Fighting' 'Misc' 'Platform' 'Puzzle' 'Racing'
 'Role-Playing' 'Shooter' 'Simulation' 'Sports' 'Strategy']


**Observations we can draw from the above output:**

> The **'u_generes'** variable stores unique features such as Action, Adventure, Fighting, Misc, Platform, Puzzle, Racing,
 Role-Playing, Shooter, Simulation, Sports, and Strategy and the total count is 12 present in the **'genre'** column of **`game_df`** dataframe.